### PR TITLE
Introduce external delegate provider to benchmark model.

### PR DESCRIPTION
**Pull request by GeorgeARM**
Allows TFLite benchmark model to load an external delegate through dlopen.

Note: This allows only a single external delegate plugin to be loaded at the moment.

### CODE REVIEW

**Comment by GeorgeARM** [comment]
@jdduke could you have a look?
As per comment, this allows a single external delegate plugin at the moment but should be good enough for initial testing.
Thank you in advance.

File: tensorflow/lite/tools/benchmark/README.md [hunk 1]
```diff
@@ -87,6 +87,13 @@ and the following optional parameters:
     `enable_op_profiling`. When this is set to true the profile of ops
     on hexagon DSP will be added to the profile table.
     Note that, the reported data on hexagon is in cycles, not in ms like on cpu.
+*   `use_external_delegate`: `bool` (default=false) \
+    Whether to use a specified external delegate library.
+*   `external_delegate_path`: `string` (default="") \
+    Path to the external delegate library.
+*   `external_delegate_options`: `string` (default="") \
+    A list of options to be passed to the external delegate library.
+    Options should be in the format of `option1:value1,option2:value2,optionN:valueN`
```
Comment (jdduke) [inline tensorflow/lite/tools/benchmark/README.md]: I'm thinking it would be better to use `;` for the separator rather than `,`?

Comment (GeorgeARM) [reply tensorflow/lite/tools/benchmark/README.md]: Yes I agree with you. Will switch Jared to using `;`

File: tensorflow/lite/tools/benchmark/external_delegate_provider.cc [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  void load(const std::string library) {
+    if (!is_loaded) {
+      void* handle = nullptr;
+      handle = dlopen(library.c_str(), RTLD_LAZY | RTLD_LOCAL);
+      if (handle == nullptr) {
+        TFLITE_LOG(INFO) << "Unable to load external delegate from : "
+                         << library;
+      } else {
+        create = reinterpret_cast<decltype(create)>(
+            dlsym(handle, "tflite_plugin_create_delegate"));
```
Comment (jdduke) [inline tensorflow/lite/tools/benchmark/external_delegate_provider.cc]: We do support non-posix platforms, so we'll either need to restrict this in such cases or use a generic intermediate which toggles between LoadLibrary/GetProcAccess

File: tensorflow/lite/tools/benchmark/external_delegate_provider.cc [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+  void LogParams(const BenchmarkParams& params) const final;
+
+  TfLiteDelegatePtr CreateTfLiteDelegate(
+      const BenchmarkParams& params) const final;
+
+  std::string GetName() const final { return "EXTERNAL"; }
+
+ private:
+  mutable ExternalLib _delegate_lib;
```
Comment (jdduke) [inline tensorflow/lite/tools/benchmark/external_delegate_provider.cc]: nit: `delegate_lib_`

File: tensorflow/lite/tools/benchmark/README.md [hunk 4]
```diff
@@ -87,6 +87,13 @@ and the following optional parameters:
     `enable_op_profiling`. When this is set to true the profile of ops
     on hexagon DSP will be added to the profile table.
     Note that, the reported data on hexagon is in cycles, not in ms like on cpu.
+*   `use_external_delegate`: `bool` (default=false) \
```
Comment (jdduke) [inline tensorflow/lite/tools/benchmark/README.md]: Should we remove this flag and just make it implicit based on `external_delegate_path`?

**Review by jdduke — CHANGES_REQUESTED** [review]
Nice!

File: tensorflow/lite/tools/benchmark/README.md (hunk 4 above)
Comment (GeorgeARM) [reply tensorflow/lite/tools/benchmark/README.md]: I think it makes sense to make it implicit, will remove.

File: tensorflow/lite/tools/benchmark/external_delegate_provider.cc (hunk 2 above)
Comment (GeorgeARM) [reply tensorflow/lite/tools/benchmark/external_delegate_provider.cc]: Can I use `tensorflow/tensorflow/core/platform/load_library.h`  ?

Comment (jdduke) [reply tensorflow/lite/tools/benchmark/external_delegate_provider.cc]: The TF platform targets aren't generally mobile portable. We can just use a lightweight shim here, e.g., https://github.com/tensorflow/tensorflow/blob/4911ca4a011c7b86087e4a420247fdc675ad8a88/tensorflow/lite/python/interpreter_wrapper/interpreter_wrapper.cc#L158. We can refactor later into a more portable helper.

**Comment by GeorgeARM** [comment]
Apologies @jdduke I overwrote the previous commit instead of pushing a patch set on top.

File: tensorflow/lite/tools/benchmark/external_delegate_provider.cc [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+// External delegate library construct
+struct ExternalLib {
+  using CreateDelegatePtr = std::add_pointer<TfLiteDelegate*(
+      const char**, const char**, size_t,
+      void (*report_error)(const char*))>::type;
+  using DestroyDelegatePtr = std::add_pointer<void(TfLiteDelegate*)>::type;
+
+  // Open a given delegate library and load the create/destroy symbols
+  void load(const std::string library) {
+    if (!is_loaded) {
```
Comment (jdduke) [inline tensorflow/lite/tools/benchmark/external_delegate_provider.cc]: Shouldn't we reload the library here? Seems like the provider could be invoked with different options?

I'd be OK simply loading ExternalLib on the stack in CreateTfLiteDelegate. Looks like we don't use Unload anywhere, and it's fine to leak the handle for benchmarking purposes.

**Review by jdduke — CHANGES_REQUESTED** [review]
Looks good, just one minor point.

File: tensorflow/lite/tools/benchmark/external_delegate_provider.cc (hunk 5 above)
Comment (GeorgeARM) [reply tensorflow/lite/tools/benchmark/external_delegate_provider.cc]: Yes I thought the same, leaking the handle should be fine for benchmarking purposes.

**Review by jdduke — DISMISSED** [review]

**Comment by gbaned** [comment]
@GeorgeARM Can you please address Ubuntu Sanity errors? Thanks!

**Comment by GeorgeARM** [comment]
> @GeorgeARM Can you please address Ubuntu Sanity errors? Thanks!

Done

**Review by jdduke — APPROVED** [review]

**Comment by GeorgeARM** [comment]
@multiverse-tf @gbaned do I have to address any issues?

**Comment by gbaned** [comment]
> @multiverse-tf @gbaned do I have to address any issues?

@GeorgeARM Nothing to do at this moment. We will come back if anything required. Thank you!

**Comment by multiverse-tf** [comment]
> > @multiverse-tf @gbaned do I have to address any issues?
> 
> @GeorgeARM Nothing to do at this moment. We will come back if anything required. Thank you!

Thank you for your contribution! Yes, We are handling the merging issue right now, and your changes will be merged

**Comment by multiverse-tf** [comment]
@GeorgeARM, a fix (https://github.com/tensorflow/tensorflow/commit/1f28da1bd5f48cf56dbddb69ca38c878d4c47776) has been submitted to address test failures of this PR. I think you might need to rebase your changes as the BUILD file has been changed to unblock the merge. Thx!

**Pull request merged**